sys/include/registry/registry.h

```diff
@@ -0,0 +1,464 @@
+ *
+ * ### Storage Facilities
+ *
+ * @ref sys_registry_store "Storage Facilities" (SF) implement the
+ * @ref registry_store_itf_t "storage interface" to allow the RIOT Registry to
+ * load, search and store configuration parameters.
+ *
+ * ## RIOT Registry usage flow
+ * - 1 The RIOT registry is initialized with @ref registry_init().
+ * - 2 Modules declare and register RHs for configuration groups by calling
+ *     @ref registry_register().
+ * - 3 SFs are registered as sources and/or destinations of configurations by
+ *     calling registry_<storage-name>_src() and registry_<storage-name>_dst().
+ *
+ * @{
+ * @file
+ *
+ * @author      Leandro Lanzieri <leandro.lanzieri@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#ifndef REGISTRY_REGISTRY_H
+#define REGISTRY_REGISTRY_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include "clist.h"
+
+/**
+ * @brief Separator character to define hierarchy in configurations names.
+ */
+#define REGISTRY_NAME_SEPARATOR    '/'
+
+/**
+ * @brief Maximum amount of levels of hierarchy in configurations names.
+ */
+#define REGISTRY_MAX_DIR_DEPTH     8
+
+/**
+ * @brief Maximum amount of characters per level in configurations names.
+ */
+#define REGISTRY_MAX_DIR_NAME_LEN  64
+
+/**
+ * @brief Maximum length of a value when converted to string
+ */
+#define REGISTRY_MAX_VAL_LEN       64
+
+/**
+ * @brief Maximum length of a configuration name.
+ * @{
+ */
+#define REGISTRY_MAX_NAME_LEN      ((REGISTRY_MAX_DIR_NAME_LEN * \
+                                    REGISTRY_MAX_DIR_DEPTH) + \
+                                    (REGISTRY_MAX_DIR_DEPTH - 1))
+/** @} */
+
+/**
+ * @brief Data types of the registry
+ *
+ * @note Float and int64 types must be enabled by defining
+ * `CONFIG_REGISTRY_USE_FLOAT` and `CONFIG_REGISTRY_USE_INT64`. Use with caution
+ * as they bloat the code size.
+ */
+typedef enum {
+    REGISTRY_TYPE_NONE = 0,  /**< No type specified */
+    REGISTRY_TYPE_INT8,      /**< 8-bits integer */
+    REGISTRY_TYPE_INT16,     /**< 16-bits integer */
+    REGISTRY_TYPE_INT32,     /**< 32-bits integer */
+    REGISTRY_TYPE_STRING,    /**< String */
+    REGISTRY_TYPE_BYTES,     /**< Binary data */
+    REGISTRY_TYPE_BOOL,      /**< Boolean */
+
+#if defined(CONFIG_REGISTRY_USE_INT64) || defined(DOXYGEN)
+    REGISTRY_TYPE_INT64,     /**< 64-bits integer */
+#endif /* CONFIG_REGISTRY_USE_INT64 */
+
+#if defined(CONFIG_REGISTRY_USE_FLOAT) || defined(DOXYGEN)
+    REGISTRY_TYPE_FLOAT,     /**< Float */
+#endif /* CONFIG_REGISTRY_USE_FLOAT */
+} registry_type_t;
+
+
+/**
+ * @brief Prototype of a callback function for the load action of a store
+ * interface
+ */
+typedef void (*load_cb_t)(char *name, char *val, void *cb_arg);
+
+/**
+ * @brief Descriptor used to check duplications in store facilities
+ */
+typedef struct {
+    char *name; /**< name of the parameter being checked */
+    char *val;  /**< value of the parameter being checked */
```

Comment:
it checks if both value or name are duplicated

Reply:
Ah I see, the name and value have to both be the same for it to be marked as a duplicate. Which means that the storage facility won't try to store it. If the name is the same but the value is different then it will store it. I guess this is a runtime optimization? Wouldn't it be better to cut out the duplicate checking and just always save over the existing entry in the storage facility, to save ROM? Also it seems like this duplicate checking would take a while for a large registry size

Reply:
Or am I missing something